### DISCUSSION TITLE
Conditional output of object only if a specified property is present in source

### DISCUSSION THREAD

**Discussion by digitalfiends** [schibsted/jslt#279] [Q&A] [answered]
First off, just wanted to say a quick thanks for this fantastic project.  I just recently discovered it, so I'm still learning what it is capable of.  Unfortunately, I've hit an issue that I can't seem to resolve and don't see an obvious solution in the tutorial/functions readmes or in the discussions here.

### Issue: 
I need a JSON object to be created only if a source key/value is present.  

### Simplified source JSON:
```
{
    "RC-INPUT-DTL-REC" : {
      "RC-ACCNT-NBR": 100,
      "RC-TTN" : 100000001
  }
}
```

### Expected output JSON (when RC-ACCNT-NBR **is** present in source)

```
{
  "Client" : {
    "ClientIdentification" : {
      "IdentificationID" : "100",
      "IdentificationCategoryText" : "Account Number"
    },
    "PersonIdentification" : {
      "IdentificationID" : "100000001"
    }
  }
}
```

### Expected output JSON (when RC-ACCNT-NBR is **not** present in source)

```
{
  "Client" : {
    "PersonIdentification" : {
      "IdentificationID" : "100000001"
    }
  }
}
```

In this case, the ClientIdentification object should not be written to the output.

### JSLT for mapping:

This is what I have for the mapping (extra stuff removed to keep things clear):
```
{ 
    "Client": {
        if(.RC-INPUT-DTL-REC.RC-ACCNT-NBR)
          "ClientIdentification": { 
              "IdentificationID": string(.RC-INPUT-DTL-REC.RC-ACCNT-NBR), 
              "IdentificationCategoryText": "Account Number" 
          }, 
        "PersonIdentification": {
            "IdentificationID": string(.RC-INPUT-DTL-REC.RC-SIN-TTN)
        }
}
```

### Result:

When RC-ACCNT-NBR is present in the source JSON, everything works well and I get the expected JSON output.  When the RC-ACCNT-NBR is not present in the source, I get the following exception:

com.schibsted.spt.data.jslt.JsltException: Object key must be string at <inline>:3:9
	at com.schibsted.spt.data.jslt.impl.PairExpression.applyKey(PairExpression.java:40)
	at com.schibsted.spt.data.jslt.impl.ObjectExpression.apply(ObjectExpression.java:91)
	at com.schibsted.spt.data.jslt.impl.PairExpression.apply(PairExpression.java:52)
	at com.schibsted.spt.data.jslt.impl.ObjectExpression.apply(ObjectExpression.java:89)
	at com.schibsted.spt.data.jslt.impl.ExpressionImpl.apply(ExpressionImpl.java:91)
	at com.schibsted.spt.data.jslt.impl.ExpressionImpl.apply(ExpressionImpl.java:73)
	at no.priv.garshol.jslt.playground.PlaygroundServer$JsltHandler.handle(PlaygroundServer.java:55)
	... snipped ...

Is it possible to achieve this with JSLT?  I would assume so and that it is just my lack of experience with the syntax.  Any help would be greatly appreciated.

Thank you!

**Comment by larsga** [comment, marked as answer]
Here's your JSLT simplified and indented, with parens added around the `if` statement, so you can see what effect it has:

```
{
  "Client": {
    ( if(.RC-INPUT-DTL-REC.RC-ACCNT-NBR) "ClientIdentification" ) : ...,
    "PersonSINIdentification": {
      "IdentificationID": string(.RC-INPUT-DTL-REC.RC-SIN-TTN)
  }
}
```

In other words, your `if` affects the `"ClientIdentification"` key, but not the value, which is always present. So when the `"RC-INPUT-DTL-REC.RC-ACCNT-NBR"` key is not in the input, the key is `null` and the value is an object, and you've produced illegal JSON.

Rewrite to the following, and it should work:

```
{
  "Client": {
    "ClientIdentification" : if(.RC-INPUT-DTL-REC.RC-ACCNT-NBR) { ... },
    "PersonSINIdentification": {
      "IdentificationID": string(.RC-INPUT-DTL-REC.RC-SIN-TTN)
  }
}
```

**Reply by digitalfiends** [reply]
Thank you, this fixed it.  I forgot that if the evaluation after `"ClientIdentification" :` returns null, it won't be included in the output JSON.  I was thinking that I had to wrap the ClientIdentification object, but as you pointed out, this results in invalid JSON.
I do have a somewhat related question around if-statements and conditionally including elements in an array.  I would love your input on that as well - I'll post it as a separate question.

Thank you again.

Edit (this is what the map ends up like, for reference):
```
{
    "Client": {
        "ClientIdentification": if(.RC-INPUT-DTL-REC.RC-OAS-ACCNT-NBR) {
            "IdentificationID": string(.RC-INPUT-DTL-REC.RC-OAS-ACCNT-NBR),
            "IdentificationCategoryText": "OAS Account Number"
        },
        "PersonIdentification": {
            "IdentificationID": string(.RC-INPUT-DTL-REC.RC-SIN-TTN)
        }
       ... snipped ...
}
```
